### PR TITLE
Allow use of ServiceProvider func to create KeyResolver for ProtectKeysWithAzureKeyVault

### PR DESCRIPTION
**Pull request by thomhurst**
Allow use of ServiceProvider func to create KeyResolver for ProtectKeysWithAzureKeyVault call (for resolving nested dependencies)

I've come across code where we want to generate our TokenCredential/KeyResolver using other services (CertificateTokenCredential which then relies on another KeyVaultClient for getting AAD certs, etc.) and without access to the ServiceProvider, this is dependency hell.

Being able to use the ServiceProvider to resolve your TokenCredential/KeyResolver allows us to retrieve registered dependencies and write cleaner setup code where needed.

### CODE REVIEW

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs [hunk 1]
```diff
@@ -54,5 +55,58 @@ public static IDataProtectionBuilder ProtectKeysWithAzureKeyVault(this IDataProt
 
             return builder;
         }
+
+        /// <summary>
+        /// Configures the data protection system to protect keys with specified key in Azure KeyVault.
+        /// </summary>
+        /// <param name="builder">The builder instance to modify.</param>
+        /// <param name="keyResolverFactory">The factory delegate to create the <see cref="IKeyEncryptionKeyResolver"/> to use for Key Vault access.</param>
```
Comment (JoshLove-msft) [inline sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs:63]: Looks like the param doc ordering got swapped.

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs [hunk 2]
```diff
@@ -54,5 +55,58 @@ public static IDataProtectionBuilder ProtectKeysWithAzureKeyVault(this IDataProt
...
+            builder.Services.AddSingleton<IConfigureOptions<KeyManagementOptions>, ConfigureKeyManagementKeyVaultEncryptorClientOptions>();
+
+            return builder;
+        }
+
+        /// <summary>
+        /// Configures the data protection system to protect keys with specified key in Azure KeyVault.
+        /// </summary>
+        /// <param name="builder">The builder instance to modify.</param>
+        /// <param name="tokenCredentialFactory">The factory delegate to create the <see cref="TokenCredential"/> to use for authenticating Key Vault access.</param>
```
Comment (JoshLove-msft) [inline sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs:89]: swapped docs ordering

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs [hunk 3]
```diff
@@ -54,5 +55,58 @@ public static IDataProtectionBuilder ProtectKeysWithAzureKeyVault(this IDataProt
 
             return builder;
         }
+
+        /// <summary>
+        /// Configures the data protection system to protect keys with specified key in Azure KeyVault.
```
Comment (heaths) [inline sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs:60]: Product name is "Key Vault".
```suggestion
        /// Configures the data protection system to protect keys with specified key in Azure Key Vault.
```

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs [hunk 4]
```diff
@@ -54,5 +55,58 @@ public static IDataProtectionBuilder ProtectKeysWithAzureKeyVault(this IDataProt
...
+                return new AzureKeyVaultXmlEncryptor(keyResolver, keyIdentifier);
+            });
+
+            builder.Services.AddSingleton<IConfigureOptions<KeyManagementOptions>, ConfigureKeyManagementKeyVaultEncryptorClientOptions>();
+
+            return builder;
+        }
+
+        /// <summary>
+        /// Configures the data protection system to protect keys with specified key in Azure KeyVault.
```
Comment (heaths) [inline sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs:86]:
```suggestion
        /// Configures the data protection system to protect keys with specified key in Azure Key Vault.
```

**Review by heaths — APPROVED** [review]

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs (hunk 1 above)
Comment (thomhurst) [reply sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs]: That was me copying and pasting from the existing method. I've reordered that one too!

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs (hunk 2 above)
Comment (thomhurst) [reply sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/AzureDataProtectionKeyVaultKeyBuilderExtensions.cs]: That was me copying and pasting from the existing method. I've reordered that one too!

**Comment by thomhurst** [comment]
@JoshLove-msft @heaths How long would something like this take to be released in a new package version?

**Comment by jsquire** [comment]
> @JoshLove-msft @heaths How long would something like this take to be released in a new package version?

Most likely, these would be released the second week of February.

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/ConfigureKeyManagementKeyVaultEncryptorClientOptions.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using Microsoft.AspNetCore.DataProtection.KeyManagement;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Options;
+
+#pragma warning disable AZC0001 //
```
Comment (jsquire) [inline sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/ConfigureKeyManagementKeyVaultEncryptorClientOptions.cs:8]: nit:  Can we please leave a short description of what is being disabled?

**Review by jsquire — APPROVED** [review]

File: sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/ConfigureKeyManagementKeyVaultEncryptorClientOptions.cs (hunk 5 above)
Comment (thomhurst) [reply sdk/extensions/Azure.Extensions.AspNetCore.DataProtection.Keys/src/ConfigureKeyManagementKeyVaultEncryptorClientOptions.cs]: Wasn't necessary - Builds without any warnings even without. Have removed

**Pull request merged**